connaisseur/image.py

```diff
@@ -27,61 +27,63 @@ class Image:
-            else "docker.io"
-        )
-        self.repository = (
-            image.removesuffix(name_tag).removeprefix(self.registry)
-        ).strip("/") or ("library" if self.registry == "docker.io" else "")
-
-        if (self.repository + self.name).lower() != self.repository + self.name:
+        digest = rf"{digest_algorithm}:{digest_hex}"
+        tag = r"[\w][\w.-]{0,127}"
+        separator = r"[_.]|__|[-]*"
+        alpha_numeric = r"[a-z0-9]+"
+        path_component = rf"{alpha_numeric}(?:{separator}{alpha_numeric})*"
+        port = r"[0-9]+"
+        domain_component = r"(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])"
+        domain_name = rf"{domain_component}(?:\.{domain_component})*"
+        ipv6 = r"\[(?:[a-fA-F0-9:]+)\]"
+        host = rf"(?:{domain_name}|{ipv6})"
+        domain = rf"{host}(?::{port})?"
+        name = rf"(?:{domain}/)?{path_component}(?:/{path_component})*"
+        reference = rf"^(?P<name>{name})(?::(?P<tag>{tag}))?(?:@(?P<digest>{digest}))?$"
+
+        match = re.search(reference, image)
+        if (not match) or (len(match.group("name")) > 255):
             msg = "{image} is not a valid image reference."
             raise InvalidImageFormatError(message=msg, image=image)
 
-    def set_digest(self, digest):
-        """
-        Set the digest to the given `digest`.
-        """
-        self.digest = digest
-
-    def has_digest(self) -> bool:
-        """
-        Return `True` if the image has a digest, `False` otherwise.
-        """
-        return self.digest is not None
+        name, tag, digest = match.groups()
+        self.digest = digest.split(":")[-1] if digest else None
+        self.tag = tag or ("latest" if not self.digest else None)
+        components = name.split("/")
+        self.name = components[-1]
+        registry_repo = components[:-1]
+
+        try:
+            registry = registry_repo[0]
+            self.registry = (
+                registry
+                if re.search(r"[.:]", registry)
+                or registry == "localhost"
+                or any(ele.isupper() for ele in registry)
+                else "docker.io"
+            )
+            self.repository = "/".join(registry_repo).removeprefix(
+                f"{self.registry}"
+            ).removeprefix("/") or ("library" if self.registry == "docker.io" else None)
+        except IndexError:
+            self.registry = "docker.io"
+            self.repository = "library"
 
     def __str__(self):
-        repo_reg = "".join(
-            f"{item}/" for item in [self.registry, self.repository] if item
-        )
-        tag = f":{self.tag}" if not self.digest else f"@sha256:{self.digest}"
-        return f"{repo_reg}{self.name}{tag}"
+        repo_reg = "/".join(item for item in [self.registry, self.repository] if item)
+        tag = f":{self.tag}" if self.tag else ""
+        digest = f"@sha256:{self.digest}" if self.digest else ""
```

Comment:
This again doesn't work if there were a different hash algo

Reply:
see above. let's stay with `sha256` for now.